### ISSUE TITLE
BUG: Linking issue in qrupdate, used by octave, on darwin

### ISSUE THREAD

**Issue by expelledboy** [NixOS/nixpkgs#107747]
**Describe the bug**

Linking issue in qrupdate, used by octave, on darwin.

Possibly effected by https://georgegarside.com/blog/macos/vector-magic-crash-el-capitan/

**To Reproduce**

`./bug.m`

```octave
data = [1,2,3;4,5,6];

X = data(:, [1, 2]);
y = data(:, 3);

[m, n] = size(X);
X = [ones(m, 1) X];

function g = sigmoid(z)
  g = 1 ./ (1 + e.^-z);
end

function [J, gradient] = costFunction(theta, X, y)
  % Number of training examples
  m = length(y);
  % Cost w/r theta
  h = sigmoid(X * theta);
  J = 1/m * (-y' * log(h) - (1-y)' * log(1-h));
  % Partial derivative w/r theta
  gradient = 1/m * X' * (h  - y);
end

% Initialize fitting parameters
initial_theta = zeros(n + 1, 1);

options = optimset('GradObj', 'on', 'MaxIter', 400);

[theta, cost] = fminunc(@(t)(costFunction(t, X, y)), initial_theta, options);
```

```sh
nix-env -iA nixpkgs.octave
octave

octave:1> bug
dyld: lazy symbol binding failed: Symbol not found: _dlartg_
  Referenced from: /nix/store/f50n1ji02mj7ilk6ksvhc8ay0ql65v0y-qrupdate-1.1.2/lib/libqrupdate.1.dylib
  Expected in: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib

dyld: Symbol not found: _dlartg_
  Referenced from: /nix/store/f50n1ji02mj7ilk6ksvhc8ay0ql65v0y-qrupdate-1.1.2/lib/libqrupdate.1.dylib
  Expected in: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
```

**Notify maintainers**

@doronbehar 

**Metadata**
```
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-darwin"`
 - host os: `Darwin 18.7.0, macOS 10.14.6`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.3.8`
 - channels(root): `"darwin, home-manager, nixpkgs-21.03pre250093.0da76dab4c2"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```
Maintainer information:
```yaml
# a list of nixos modules affected by the problem
module:
  octave
  qrupdate
```

**Comment by expelledboy** [comment]
I am not sure what libBLAS is, but I grok'd and found a package which includes [Darwin package install for Accelerate Framework Library](https://github.com/khssnv/nixpkgs/blob/c1e60d958402a7750193c803e5e52fc0d7017b05/pkgs/applications/gis/gmt/default.nix), might be useful?

**Comment by expelledboy** [comment]
Or perhaps the `DYLD_LIBRARY_PATH` [needs to be set](https://github.com/maxhbr/myconfig/blob/beccc13e292450a382040a20f84a873a146fca35/nixpkgs/pkgs/development/libraries/science/math/itpp/default.nix)? Also another approach where they [copy the libs out](https://discourse.julialang.org/t/using-julia-with-nixos/35129/24) during build.

Hope this helps cause I am at a loss myself what is causing this, or what I have to change.

**Comment by doronbehar** [comment]
So according to https://georgegarside.com/blog/macos/vector-magic-crash-el-capitan/, `Accelerate` can be used instead of our `blas`? I once talked with `@danieldk` (not ccing him) about this when I updated suitesparse at https://github.com/NixOS/nixpkgs/pull/98485/files#r493217330 .

Maybe the problem is that we add `Accelerate` on Darwin: https://github.com/NixOS/nixpkgs/blob/c7f75838c360473805afcf5fb2fa65e678efd94b/pkgs/development/interpreters/octave/default.nix#L143 but according to the policy implied in that conversation, we wish to not use `Accelerate`. So, perhaps: you can try and remove it from the build?

**Comment by doronbehar** [comment]
I mean that our policy should be consistent - we should either use our blas or Accelerate's blas and currently the situation seems a bit self inconsistent.

**Comment by expelledboy** [comment]
What would be the pros of sticking with the Darwin Accelerate? The thing is that there are afew related packages that are using OpenBLAS, not sure it matters much though with nixpkgs.

How can I help?

**Comment by doronbehar** [comment]
> What would be the pros of sticking with the Darwin Accelerate? The thing is that there are afew related packages that are using OpenBLAS, not sure it matters much though with nixpkgs.
> 
> How can I help?

Good question. As [I said previously](https://github.com/NixOS/nixpkgs/issues/107747#issuecomment-751519992), you can test the octave build without Accelerate, and see if it even builds. If it does, you can test that bug with that code snippet, and if removing Accelerate solves the bug for you, you can open a PR.

If that will not work, we can ask for help from `@danieldk`.

**Comment by expelledboy** [comment]
~/.config/nixpkgs/overlays/octave.nix

```nix
self: super:
let
  inherit (super) lib;
in
{
  octave = super.octave.overrideAttrs (o: {
    buildInputs = (lib.remove self.darwin.apple_sdk.frameworks.Accelerate o.buildInputs);
  });
}
```

Okay I cant actually tell if this was applied, but it re-build everything so I am assuming. The problem still persists. Do I not need to change the lib path during build?

**Comment by doronbehar** [comment]
> Okay I cant actually tell if this was applied, but it re-build everything so I am assuming. The problem still persists. Do I not need to change the lib path during build?

I too think you got that override applied. IIRC you can verify this by looking at the `buildInputs` key of:

```sh
nix show-derivation -f nixpkgs octave
```

Another option would be to use a nixpkgs clone and remove that dep from the expression, and build it with: `nix-build -A octave`.

**Comment by expelledboy** [comment]
That easier then I expected, but problem still persists.

**Comment by doronbehar** [comment]
@danieldk perhaps you could share your point of view regarding our blas vs darwin's `Accelerate`'s version of blas?

**Comment by danieldk** [comment]
> @danieldk perhaps you could share your point of view regarding our blas vs darwin's `Accelerate`'s version of blas?

We should avoid specific BLAS/LAPACK implementations in derivations, because we have a mechanism that makes it easy to override what implementation is used. See:

https://nixos.org/manual/nixpkgs/stable/#sec-overlays-alternatives-blas-lapack

If derviations use a specific BLAS or LAPACK implementation, then overriding them becomes harder.

Regardless of the override mechanism, one would normally want to use Accelerate on macOS, since it is the most-optimized BLAS/LAPACK implementation on Apple hardware. Though there are some limitations -- if I recall correctly, Accelerate does not provide the LAPACKE interface, so a LAPACKE -> LAPACK shim is needed.

Since I don't have a Mac anymore, I am afraid I am of little help here. The only thing I could think of is that the linker is searching  the wrong library. In your report you mention `/System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib` could you check if there is a `libLAPACK.dylib` in the same directory. If so, you could use `nm` to inspect if the missing symbol is in `libLAPACK.dylib`. If it has `dlartg_`, the problem is possibly that the program should be linked against `libLAPACK.dylib` instead (since dlartg is a LAPACK function).

**Comment by smancill** [comment]
I've been working on building [Shogun](pkgs/applications/science/machine-learning/shogun/default.nix) on Darwin, and I just encountered this same problem. Shogun is not using any framework as build input.

```
dyld: lazy symbol binding failed: Symbol not found: _dsygvx_
  Referenced from: /private/tmp/nix-build-shogun-6.1.4.drv-0/source/build/src/shogun/libshogun.18.dylib
  Expected in: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
```

The problem is that `libshogun.18.dylib` is using the wrong name for the `libblas.dylib` dependency (it should be a full path to the nix store input):

```
$ otool -L /private/tmp/nix-build-shogun-6.1.4.drv-0/source/build/src/shogun/libshogun.18.dylib
/private/tmp/nix-build-shogun-6.1.4.drv-0/source/build/src/shogun/libshogun.18.dylib:
        @rpath/libshogun.18.dylib (compatibility version 18.0.0, current version 18.0.0)
        libblas.dylib (compatibility version 0.0.0, current version 0.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.60.2)
        /nix/store/xp5dhzzkhnvqy8ra7p9c3a4x40xn0sn9-openblas-0.3.15/lib/libopenblas.0.dylib (compatibility version 0.0.0, current version 0.0.0)
        /nix/store/5np1akalmmqn54bzkzyd6s9mdqa0ar20-glpk-4.65/lib/libglpk.40.dylib (compatibility version 44.0.0, current version 44.0.0)
        /nix/store/00zpf7n02qndqbvqvnqfighy3q4yn5wd-protobuf-3.17.3/lib/libprotobuf.28.dylib (compatibility version 29.0.0, current version 29.3.0)
        /nix/store/l4jdxbxa5wqvyy3wmvrx4rv1izfwm8p9-json-c-0.15/lib/libjson-c.5.dylib (compatibility version 5.0.0, current version 5.1.0)
        /nix/store/4xp86qi542rlddjxndff141l4cyqw9ri-libxml2-2.9.12/lib/libxml2.2.dylib (compatibility version 12.0.0, current version 12.12.0)
        /nix/store/a4818rc7lrah3491qm52rz3rz7a1hrzf-hdf5-1.12.1/lib/libhdf5.200.dylib (compatibility version 202.0.0, current version 202.0.0)
        /nix/store/x87bqc42xmskjwf5c2a9bbzkd8aqwrca-zlib-1.2.11/lib/libz.dylib (compatibility version 1.0.0, current version 1.2.11)
        /nix/store/b2wj786d4qizfamzwhc6qfsb1yaaqsm5-curl-7.76.1/lib/libcurl.4.dylib (compatibility version 12.0.0, current version 12.0.0)
        /nix/store/5f1qxwxxp9mag9r91mdfs3w1x3y99kxm-bzip2-1.0.6.0.2/lib/libbz2.1.dylib (compatibility version 2.0.0, current version 2.6.0)
        /nix/store/mq766n41v9djm2rf74dpg4hd7n5mbwwh-xz-5.2.5/lib/liblzma.5.dylib (compatibility version 8.0.0, current version 8.5.0)
        /nix/store/3rzb7cy2qnhl99fws7qxh3brm82afy2f-snappy-1.1.8/lib/libsnappy.1.dylib (compatibility version 1.0.0, current version 1.1.8)
        /nix/store/5cv4vf1jybf00q4n82ys54pml0gvhr7d-lzo-2.10/lib/liblzo2.2.dylib (compatibility version 3.0.0, current version 3.0.0)
        /nix/store/7cbfgb1wrdmyai3zp51f5iq6cf3afljx-nlopt-2.7.0/lib/libnlopt.0.dylib (compatibility version 0.0.0, current version 0.11.0)
        /nix/store/zzrpnbmpwgsfrsyz2lfgvha6prn5mz4r-lp_solve-5.5.2.11/lib/liblpsolve55.dylib (compatibility version 5.5.0, current version 5.5.0)
        /nix/store/65lfbr911pb98qhzi5414hx17b5j7myw-ColPack-1.0.10/lib/libColPack.0.dylib (compatibility version 1.0.0, current version 1.0.0)
        /nix/store/fsz5rxl37pamwhqkmp9w4jryfc4pfky3-libcxx-7.1.0/lib/libc++.1.0.dylib (compatibility version 1.0.0, current version 1.0.0)
```

But the linking command is using the proper `libblas.dylib` library:

```
/nix/store/jx31ah14p6pf488i7wwd88y2izmxi0wj-clang-wrapper-7.1.0/bin/clang++
  -Wall
  -Wno-unused-parameter
  -Wformat
  -Wformat-security
  -Wparentheses
  -Wshadow
  -Wno-unknown-pragmas
  -Wno-deprecated
  -Qunused-arguments
  -fcolor-diagnostics
  -O3
  -funroll-loops
  -dynamiclib
  -Wl,-headerpad_max_install_names
  -compatibility_version 18.0.0
  -current_version 18.0.0
  -o libshogun.18.0.dylib
  -install_name @rpath/libshogun.18.dylib
  <object_files>
  /nix/store/xvradv9zllldhlpyk93lyylm051222r9-blas-3/lib/libblas.dylib
  -lm
  -ldl
  /nix/store/fkwk7lszvf42ilk09p5z0vvb6sy0hkkq-lapack-3/lib/liblapacke.dylib
  /nix/store/5np1akalmmqn54bzkzyd6s9mdqa0ar20-glpk-4.65/lib/libglpk.dylib
  /nix/store/00zpf7n02qndqbvqvnqfighy3q4yn5wd-protobuf-3.17.3/lib/libprotobuf.dylib
  /nix/store/l4jdxbxa5wqvyy3wmvrx4rv1izfwm8p9-json-c-0.15/lib/libjson-c.dylib
  /nix/store/4xp86qi542rlddjxndff141l4cyqw9ri-libxml2-2.9.12/lib/libxml2.dylib
  /nix/store/a4818rc7lrah3491qm52rz3rz7a1hrzf-hdf5-1.12.1/lib/libhdf5.dylib
  /nix/store/x87bqc42xmskjwf5c2a9bbzkd8aqwrca-zlib-1.2.11/lib/libz.dylib
  /nix/store/4npah0kdy1jgqsbb4sw2vpbv9l2zsh18-Libsystem-1238.60.2/lib/libdl.tbd
  /nix/store/4npah0kdy1jgqsbb4sw2vpbv9l2zsh18-Libsystem-1238.60.2/lib/libm.tbd
  /nix/store/b2wj786d4qizfamzwhc6qfsb1yaaqsm5-curl-7.76.1/lib/libcurl.dylib
  /nix/store/x87bqc42xmskjwf5c2a9bbzkd8aqwrca-zlib-1.2.11/lib/libz.dylib
  /nix/store/5f1qxwxxp9mag9r91mdfs3w1x3y99kxm-bzip2-1.0.6.0.2/lib/libbz2.dylib
  /nix/store/mq766n41v9djm2rf74dpg4hd7n5mbwwh-xz-5.2.5/lib/liblzma.dylib
  /nix/store/3rzb7cy2qnhl99fws7qxh3brm82afy2f-snappy-1.1.8/lib/libsnappy.dylib
  /nix/store/5cv4vf1jybf00q4n82ys54pml0gvhr7d-lzo-2.10/lib/liblzo2.dylib
  /nix/store/7cbfgb1wrdmyai3zp51f5iq6cf3afljx-nlopt-2.7.0/lib/libnlopt.dylib
  /nix/store/zzrpnbmpwgsfrsyz2lfgvha6prn5mz4r-lp_solve-5.5.2.11/lib/liblpsolve55.dylib
  /nix/store/65lfbr911pb98qhzi5414hx17b5j7myw-ColPack-1.0.10/lib/libColPack.dylib
  /nix/store/4npah0kdy1jgqsbb4sw2vpbv9l2zsh18-Libsystem-1238.60.2/lib/libdl.tbd
  /nix/store/4npah0kdy1jgqsbb4sw2vpbv9l2zsh18-Libsystem-1238.60.2/lib/libm.tbd
  /nix/store/b2wj786d4qizfamzwhc6qfsb1yaaqsm5-curl-7.76.1/lib/libcurl.dylib
  /nix/store/5f1qxwxxp9mag9r91mdfs3w1x3y99kxm-bzip2-1.0.6.0.2/lib/libbz2.dylib
  /nix/store/mq766n41v9djm2rf74dpg4hd7n5mbwwh-xz-5.2.5/lib/liblzma.dylib
  /nix/store/3rzb7cy2qnhl99fws7qxh3brm82afy2f-snappy-1.1.8/lib/libsnappy.dylib
  /nix/store/5cv4vf1jybf00q4n82ys54pml0gvhr7d-lzo-2.10/lib/liblzo2.dylib
  /nix/store/7cbfgb1wrdmyai3zp51f5iq6cf3afljx-nlopt-2.7.0/lib/libnlopt.dylib
  /nix/store/zzrpnbmpwgsfrsyz2lfgvha6prn5mz4r-lp_solve-5.5.2.11/lib/liblpsolve55.dylib
  /nix/store/65lfbr911pb98qhzi5414hx17b5j7myw-ColPack-1.0.10/lib/libColPack.dylib
```

It seems an issue with the linker wrapper that prefers the system `libblas.dylib` but I don't know that much.

I've solved it by changing the name of the dependency:

``` nix
postBuild = let                                                                     
  fix_libblas = "install_name_tool -change libblas.dylib ${blas}/lib/libblas.dylib";
in lib.optional stdenv.isDarwin ''                                                  
  ${fix_libblas} src/shogun/libshogun.dylib                                         
                                                                                    
  # Fix tests                                                                       
  ${fix_libblas} bin/shogun-unit-test                                               
  # a bunch of other binaries            
'';                                                                                 
```

**Comment by smancill** [comment]
Should have been fixed by #135493 and #136535.
